### ISSUE TITLE
union types in nested objects causes error [BUG]

### ISSUE THREAD

**Issue by julada** [dynamoose/dynamoose#1100]
### Summary:

I tried to create a schema with union types in a nested object.

#### Schema
```
const NestedObjectWithUnionType = {
  type: Object,
  schema: {
    attribute1: [String, Boolean, Number],
    attribute2: String,
  },
};

const MyModel = dynamoose.model(
  `test`,
  new dynamoose.Schema({
    exampleId: {
      type: String,
      hashKey: true,
    },
    name: String,
    myNestedObject: NestedObjectWithUnionType,
  })
);
```

### Current output and behavior (including stack trace):

    InvalidParameter: Index must be at root object and not nested in object or array.
      at node_modules/dynamoose/lib/Schema.ts:587:11
          at Array.forEach (<anonymous>)
      at new Schema (node_modules/dynamoose/lib/Schema.ts:585:21)
      at Object.<anonymous> (src/models/article-mapping.model.ts:648:102)
      at Object.<anonymous> (src/modules/mapping/validate/validate-mapping.ts:1456:27)

### Environment:

Operating System: macOS
Operating System Version: Sierra
Node.js version (`node -v`): 14.8.0
NPM version: (`npm -v`): 6.14.5
Dynamoose version: 2.5.0

**Comment by ipekmuhammet** [comment]
You have to define types and schemas of sub documents, if its not primitive

const NestedObjectWithUnionType = {
  type: Object,
  schema: {
    attribute1: {
      type: Array,
      schema: [String, Boolean, Number]
    },
    attribute2: String,
  },
};

**Comment by fishcharlie** [comment]
Thank you for your issue. This project is successful due to the reports and help we get from the community in improving the project.

This project has a lot of guidelines that help ensure the project continues to run smoothly and project maintainers aren't burdened with a lot of back and forth discussion in order to advance the project and resolve issues.

**Sadly it has been determined that your issue does not follow those guidelines.**

This could be due to the following *checked* items (unchecked items were satisfied, found irrelevant, or not checked):

General

- [x] Issue template not filled out
- [ ] Issue is unclear
- [ ] Issue doesn't explain what has already been tried and what attempts you have made to resolve the issue

Example (MCVE)

- [ ] Example provided has too much irrelevant code and is not considered minimal (*Minimal*)
- [ ] Example provided has undefined variables that make it impossible to reproduce (*Complete*)
- [ ] Example provided is missing code necessary to reproduce the error (*Complete*)
- [x] When trying to reproduce example provided I did not receive the same error message as described in issue (*Verifiable*)
- [ ] No example provided (*Example*)

Error

- [ ] No error included
- [ ] Error included is unclear or confusing
- [ ] No stack trace included in error
- [ ] No information about current and expected result

Read through our [contributing guidelines](https://github.com/dynamoose/dynamoose/blob/master/CONTRIBUTING.md) for more information about what is expected when submitting issues.

If you have any questions about why this message was posted on your issue, please comment below with your specific question and we'd be happy to discuss with you further. If you believe you have followed the guidelines and still received this error message, please comment your reasoning below for why you believe you have followed the guidelines checked above, and we'll take another look at it.

While we understand that these guidelines create a larger burden on you, it is critical you follow them. The project maintainers have to handle everyone's issues while you only have to write yours, so it is imperative you make our jobs easier and help save us time by following the guidelines.

In order to proceed, please create a *new issue* ensuring you address all guidelines checked above (while also taking note to not violate any guidelines you followed).

---

Additional Information:

- Code provided doesn't result in error message shown in original post.

---

Disclosure: This was based on a [GitHub Saved Reply](https://docs.github.com/en/free-pro-team@latest/github/writing-on-github/working-with-saved-replies). The checkboxes were manually checked by myself based on your original issue.

**Comment by julada** [comment]
> You have to define types and schemas of sub documents, if its not primitive
> 
> ```
> const NestedObjectWithUnionType = {
>   type: Object,
>   schema: {
>     attribute1: {
>       type: Array,
>       schema: [String, Boolean, Number]
>     },
>     attribute2: String,
>   },
> };
> ```

I don't want an array... 
I just want to allow a boolean OR string in a nested attribute. 

@fishcharlie which information is missing? I provided everything to reproduce the error. The documentation is unclear at this point...

**Comment by fishcharlie** [comment]
@julada I ran the following code (just like in your original post).

```js
const dynamoose = require("dynamoose");
dynamoose.aws.ddb.local();

const NestedObjectWithUnionType = {
	type: Object,
	schema: {
	  attribute1: [String, Boolean, Number],
	  attribute2: String,
	},
  };

  const MyModel = dynamoose.model(
	`test`,
	new dynamoose.Schema({
	  exampleId: {
		type: String,
		hashKey: true,
	  },
	  name: String,
	  myNestedObject: NestedObjectWithUnionType,
	})
  );
```

And got this as the result:

```
charliefish tmp4 % node index.js
charliefish tmp4 % 
```

As you can see, no error, no problem. It ran successfully.

**Comment by julada** [comment]
Mh.. I'll try tomorrow again. I received this error in a larger schema and was able to reproduce it using the example given. Maybe there's a difference between local db and real dynamo?

**Comment by julada** [comment]
> @julada I ran the following code (just like in your original post).
> 
> ```js
> const dynamoose = require("dynamoose");
> dynamoose.aws.ddb.local();
> 
> const NestedObjectWithUnionType = {
> 	type: Object,
> 	schema: {
> 	  attribute1: [String, Boolean, Number],
> 	  attribute2: String,
> 	},
>   };
> 
>   const MyModel = dynamoose.model(
> 	`test`,
> 	new dynamoose.Schema({
> 	  exampleId: {
> 		type: String,
> 		hashKey: true,
> 	  },
> 	  name: String,
> 	  myNestedObject: NestedObjectWithUnionType,
> 	})
>   );
> ```
> 
> And got this as the result:
> 
> ```
> charliefish tmp4 % node index.js
> charliefish tmp4 % 
> ```
> 
> As you can see, no error, no problem. It ran successfully.

I tried exactly the same code and got the same error message...

**Comment by fishcharlie** [comment]
@julada Not sure what you tell you. Attached is a screen recording showing my result. As you can see there is no error. Exact same code.

https://user-images.githubusercontent.com/860375/106070703-34c24180-60c2-11eb-8949-b64cb7ed4f6d.mp4

**Comment by julada** [comment]
@fishcharlie 
I'm confused.
https://user-images.githubusercontent.com/9818404/106119072-12c5cf00-6155-11eb-94b3-9d5feb582e80.mov

**Comment by fishcharlie** [comment]
@julada Can you zip up that entire `dynamoose-bug` folder and post it here?

**Comment by fishcharlie** [comment]
@julada Oh maybe the problem is that this was fixed in #985. Just hasn't been released yet. Still doesn't completely explain why I couldn't reproduce it. Maybe NPM was caching a local version?? A new release with this fix should be coming out soon.
